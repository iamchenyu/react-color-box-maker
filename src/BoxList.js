import React, { useState } from "react";
import { v4 as uuid } from "uuid";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import "./BoxList.css";

const BoxList = () => {
  const INITIAL_STATE = [
    { id: uuid(), color: "DarkSeaGreen", width: "150px", height: "150px" },
    { id: uuid(), color: "LavenderBlush", width: "100px", height: "200px" },
    { id: uuid(), color: "Lavender", width: "300px", height: "200px" },
    { id: uuid(), color: "LightSalmon", width: "200px", height: "100px" },
  ];

  const [boxes, setBoxes] = useState(INITIAL_STATE);

  const removeBoxHandler = (removeId) => {
    setBoxes(boxes.filter((box) => box.id !== removeId));
  };

  const addBoxHandler = ({ color, width, height }) => {
    width = width + "px";
    height = height + "px";
    setBoxes((boxes) => [...boxes, { id: uuid(), color, width, height }]);
  };

  const renderBox = boxes.map(({ id, color, width, height }) => (
    <Box
      key={id}
      id={id}
      color={color}
      width={width}
      height={height}
      removeBox={removeBoxHandler}
    />
  ));

  return (
    <>
      <NewBoxForm addBox={addBoxHandler} />
      <div className="BoxList">{renderBox}</div>
    </>
  );
};

export default BoxList;
