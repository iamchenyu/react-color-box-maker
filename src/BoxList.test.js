import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

it("renders without crashing", () => {
  render(<BoxList />);
});

it("matches the snapshot", () => {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("removes the box after click the button", () => {
  const { queryByTestId } = render(<BoxList />);
  expect(queryByTestId("Lavender")).toBeInTheDocument();

  fireEvent.click(queryByTestId("Lavender-button"));

  expect(queryByTestId("Lavender")).not.toBeInTheDocument();
});

describe("tests for the form", () => {
  it("works: adds a new box", () => {
    const { queryByTestId, queryByLabelText } = render(<BoxList />);
    expect(queryByTestId("#6b8e23")).not.toBeInTheDocument();

    const colorInput = queryByLabelText("Color");
    const widthInput = queryByLabelText("Width");
    const heightInput = queryByLabelText("Height");
    const submitButton = queryByTestId("submit-form-button");

    fireEvent.change(colorInput, { target: { value: "#6b8e23" } });
    fireEvent.change(widthInput, { target: { value: "200" } });
    fireEvent.change(heightInput, { target: { value: "200" } });
    fireEvent.click(submitButton);

    expect(queryByTestId("#6b8e23")).toBeInTheDocument();
    expect(queryByTestId("#6b8e23")).toHaveStyle({
      backgroundColor: "#6b8e23",
      width: "200px",
      height: "200px",
    });
  });

  it("fails: invalid width/height", () => {
    const { queryByTestId, queryByLabelText } = render(<BoxList />);
    expect(queryByTestId("#db7093")).not.toBeInTheDocument();
    expect(queryByTestId("invalid-message")).not.toBeInTheDocument();

    const colorInput = queryByLabelText("Color");
    const widthInput = queryByLabelText("Width");
    const submitButton = queryByTestId("submit-form-button");

    fireEvent.change(colorInput, { target: { value: "#6b8e23" } });
    fireEvent.change(widthInput, { target: { value: "-200" } });
    fireEvent.click(submitButton);
    expect(queryByTestId("invalid-message")).toBeInTheDocument();
    expect(queryByTestId("#db7093")).not.toBeInTheDocument();
  });
});
