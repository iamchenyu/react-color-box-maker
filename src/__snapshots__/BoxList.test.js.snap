// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <form
    class="NewBoxForm"
  >
    <div
      class="NewBoxForm-input-group"
    >
      <label
        for="color"
      >
        Color
      </label>
      <input
        id="color"
        name="color"
        type="color"
        value="#7B68EE"
      />
    </div>
    <div
      class="NewBoxForm-input-group"
    >
      <label
        for="width"
      >
        Width
      </label>
      <input
        id="width"
        name="width"
        type="number"
        value="200"
      />
      <span>
        px
      </span>
    </div>
    <div
      class="NewBoxForm-input-group"
    >
      <label
        for="height"
      >
        Height
      </label>
      <input
        id="height"
        name="height"
        type="number"
        value="200"
      />
      <span>
        px
      </span>
    </div>
    <button
      data-testid="submit-form-button"
    >
      Create a box
    </button>
  </form>
  <div
    class="BoxList"
  >
    <div
      class="Box"
      data-testid="DarkSeaGreen"
      style="background-color: DarkSeaGreen; width: 150px; height: 150px;"
    >
      <button
        class="Box-button"
        data-testid="DarkSeaGreen-button"
        role="button"
      >
        X
      </button>
    </div>
    <div
      class="Box"
      data-testid="LavenderBlush"
      style="background-color: LavenderBlush; width: 100px; height: 200px;"
    >
      <button
        class="Box-button"
        data-testid="LavenderBlush-button"
        role="button"
      >
        X
      </button>
    </div>
    <div
      class="Box"
      data-testid="Lavender"
      style="background-color: Lavender; width: 300px; height: 200px;"
    >
      <button
        class="Box-button"
        data-testid="Lavender-button"
        role="button"
      >
        X
      </button>
    </div>
    <div
      class="Box"
      data-testid="LightSalmon"
      style="background-color: LightSalmon; width: 200px; height: 100px;"
    >
      <button
        class="Box-button"
        data-testid="LightSalmon-button"
        role="button"
      >
        X
      </button>
    </div>
  </div>
</DocumentFragment>
`;
